log_entry.go

```diff
@@ -0,0 +1,242 @@
+	return entry.AsLevel(WarnLevel)
+}
+
+// AsError clones the entry into a new log entry and sets the level to `error`
+// Make sure you call this method before calling WithField, WithFields and WithError methods
+func (entry *LogEntry) AsError() *LogEntry {
+	return entry.AsLevel(ErrorLevel)
+}
+
+// AsFatal clones the entry into a new log entry and sets the level to `fatal`
+// Make sure you call this method before calling WithField, WithFields and WithError methods
+func (entry *LogEntry) AsFatal() *LogEntry {
+	return entry.AsLevel(FatalLevel)
+}
+
+// AsPanic clones the entry into a new log entry and sets the level to `panic`
+// Make sure you call this method before calling WithField, WithFields and WithError methods
+func (entry *LogEntry) AsPanic() *LogEntry {
+	return entry.AsLevel(PanicLevel)
+}
+
+// WithField adds a field to the log entry, note that it doesn't log until you call Write.
+func (entry *LogEntry) WithField(key string, value interface{}) *LogEntry {
+	if entry.Level > entry.Logger.level() {
+		return entry
+	}
+	//Do not change this to Fields{key:value}. You will end up getting more allocations
+	fields := make(Fields, 1)
+	fields[key] = value
+	return entry.WithFields(fields)
+}
+
+// WithFields adds a struct of fields to the log entry
+func (entry *LogEntry) WithFields(fields Fields) *LogEntry {
+	if entry.Level > entry.Logger.level() {
+		return entry
+	}
+	data := make(Fields, len(entry.Data)+len(fields))
+	for k, v := range entry.Data {
+		data[k] = v
+	}
+	for k, v := range fields {
+		data[k] = v
+	}
+	return &LogEntry{
+		Data:   data,
+		Level:  entry.Level,
+		Logger: entry.Logger,
+	}
+}
+
+// WithError adds an error as single field to the log entry
+func (entry *LogEntry) WithError(err error) *LogEntry {
+	return entry.WithField(ErrorKey, err)
+}
+
+func (entry *LogEntry) Writef(format string, args ...interface{}) {
+	entry.write(formatted, format, args...)
+}
+
+func (entry *LogEntry) Write(args ...interface{}) {
+	entry.write(unformatted, "", args...)
+}
+
+func (entry *LogEntry) Writeln(args ...interface{}) {
+	entry.write(newLine, "", args...)
+}
+
+func (entry *LogEntry) write(mode formatMode, format string, args ...interface{}) {
+	if entry.Logger.level() >= entry.Level {
+		message := constructMessage(mode, format, args...)
+		entry.log(entry.Level, message)
+	}
+}
+
+func constructMessage(mode formatMode, format string, args ...interface{}) string {
+	switch mode {
+	case formatted:
+		return fmt.Sprintf(format, args...)
+	case unformatted:
+		return fmt.Sprint(args...)
+	case newLine:
+		return sprintlnn(args...)
+	}
+	return fmt.Sprintf(format, args...)
+}
+
+// This function is not declared with a pointer value because otherwise
+// race conditions will occur when using multiple goroutines
+func (entry LogEntry) log(level Level, msg string) {
+	var buffer *bytes.Buffer
+	entry.Time = time.Now()
+	entry.Level = level
+	entry.Message = msg
+
+	entry.Logger.mu.Lock()
```

Comment:
what happens if one of the hook panic?

Reply:
You are right. We should recover. Will fix it in the next commit.